### ISSUE TITLE
gsl recipe failing

### ISSUE THREAD

**Issue by bjodah** [conda/conda-recipes#400]
I don't know why it insists on downloading a version of gcc, many recipes requires compilers without them being specified as a requirement, right?

```
bjorn@urania:~/vc/conda-recipes$ conda build gsl
Removing old build environment
Removing old work directory
BUILD START: gsl-1.16-1
Fetching package metadata: ......
Solving package specifications: .
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    gmp-5.1.2                  |                2         594 KB  defaults
    isl-0.12.2                 |                0         1.1 MB  defaults
    mpfr-3.1.2                 |                0         407 KB  defaults
    cloog-0.18.0               |                0         617 KB  defaults
    mpc-1.0.1                  |                0          61 KB  defaults
    gcc-4.8.2                  |                2       251.2 MB  defaults
    ------------------------------------------------------------
                                           Total:       254.0 MB

The following NEW packages will be INSTALLED:

    cloog: 0.18.0-0 defaults
    gcc:   4.8.2-2  defaults
    gmp:   5.1.2-2  defaults
    isl:   0.12.2-0 defaults
    mpc:   1.0.1-0  defaults
    mpfr:  3.1.2-0  defaults

Fetching packages ...
gmp-5.1.2-2.ta 100% |####################################################################################################################################################################| Time: 0:00:00 921.03 kB/s
isl-0.12.2-0.t 100% |####################################################################################################################################################################| Time: 0:00:00   1.31 MB/s
mpfr-3.1.2-0.t 100% |####################################################################################################################################################################| Time: 0:00:00 783.55 kB/s
cloog-0.18.0-0 100% |####################################################################################################################################################################| Time: 0:00:00 987.18 kB/s
mpc-1.0.1-0.ta 100% |####################################################################################################################################################################| Time: 0:00:00 288.19 kB/s
gcc-4.8.2-2.ta 100% |####################################################################################################################################################################| Time: 0:00:39   6.70 MB/s
Extracting packages ...
[      COMPLETE      ]|#######################################################################################################################################################################################| 100%
Linking packages ...
[      COMPLETE      ]|#######################################################################################################################################################################################| 100%
Removing old work directory
Source cache directory is: /home/bjorn/miniconda/conda-bld/src_cache
Downloading source to cache: gsl-1.16.tar.gz
Downloading http://ftp.club.cc.cmu.edu/pub/gnu/gsl/gsl-1.16.tar.gz
gsl-1.16.tar.g 100% |####################################################################################################################################################################| Time: 0:00:01   2.71 MB/s
Success
Extracting download
Package: gsl-1.16-1
source tree in: /home/bjorn/miniconda/conda-bld/work/gsl-1.16
+ ./configure --prefix=/home/bjorn/miniconda/envs/_build
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking whether to enable maintainer-specific portions of Makefiles... no
checking for a sed that does not truncate output... /bin/sed
checking whether make sets $(MAKE)... (cached) yes
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/home/bjorn/miniconda/conda-bld/work/gsl-1.16':
configure: error: C compiler cannot create executables
See `config.log' for more details
Command failed: /bin/bash -x -e /home/bjorn/vc/conda-recipes/gsl/build.sh
```

**Comment by stuarteberg** [comment]
>I don't know why it insists on downloading a version of gcc, many recipes requires compilers without them being specified as a requirement, right?

For libraries written in C (not C++), it's true that the version of gcc doesn't really matter.  Obtaining `gcc` via conda (instead of your OS) is one option available to the package author, and ideally an easy one.

Regarding the failure you posted:

```
checking for gcc... gcc
checking whether the C compiler works... no
configure: error: in `/home/bjorn/miniconda/conda-bld/work/gsl-1.16':
configure: error: C compiler cannot create executables
See `config.log' for more details
```

I would like to see the contents of `config.log`, so I figure out why the `gcc` package isn't working in this case.  Can you post it?

**Comment by stuarteberg** [comment]
Actually, I think I know why the recipe is failing for you.  Please add the `asmeurer` channel to your `~/.condarc` file, and modify the recipe requirement so it specifies an exact gcc version: `gcc==4.8.5`. (The 4.8.2 package in conda doesn't work properly on most versions of Linux, but the 4.8.5 version does.)

**Comment by bjodah** [comment]
@stuarteberg thank you for your help. asmeurer's version of gcc got pulled as soon as I added him asa channel, but now there is another error:
```
bjorn@urania:~/vc/conda-recipes$ conda build gsl
Removing old build environment
Removing old work directory
BUILD START: gsl-1.16-1
Fetching package metadata: ........
Solving package specifications: .
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    gcc-4.8.5                  |                0       252.5 MB  asmeurer

The following NEW packages will be INSTALLED:

    cloog: 0.18.0-0 asmeurer
    gcc:   4.8.5-0  asmeurer
    gmp:   5.1.2-2  asmeurer
    isl:   0.12.2-0 asmeurer
    mpc:   1.0.1-0  asmeurer
    mpfr:  3.1.2-0  asmeurer

Fetching packages ...
gcc-4.8.5-0.ta 100% |#####################################################################################################| Time: 0:00:24  10.60 MB/s
Extracting packages ...
[      COMPLETE      ]|########################################################################################################################| 100%
Linking packages ...
[mpc                 ]|################################################################################                     gcc: fatal error: braced spec ‘%:sanitize(address):%{!shared:libasan_preinit%O%s} %{static-libasan:%{!shared:-Bstatic --whole-archive -lasan --no-whole-archive -Bdynamic}}%{!static-libasan:-lasan}}     %{%:sanitize(thread):%{static-libtsan:%{!shared:-Bstatic --whole-archive -ltsan --no-whole-archive -Bdynamic}}%{!static-libtsan:-ltsan}}     %{%:sanitize(leak):%{static-liblsan:%{!shared:-Bstatic --whole-archive -llsan --no-whole-archive -Bdynamic}}%{!static-liblsan:-llsan}}’ is invalid at ‘%’
compilation terminated.
Installation failed: gcc is not able to compile a simple 'Hello, World' program.
Error: post-link failed for: gcc-4.8.5-0###################################################################################                    |  83%
```

I am unable to find any *.log files in my conda root directory. Do I need to enbale debugging somehow?

**Comment by stuarteberg** [comment]
What OS are you using?

**Comment by bjodah** [comment]
Linux (Ubuntu 14.04)

**Comment by stuarteberg** [comment]
In your conda installation, there is a file named `pkgs/gcc-4.8.5-0/bin/.gcc-post-link.sh`.  The last portion of that file runs a quick test to see if the package is working properly.  Comment out that portion of the file, and then install gcc to a new environment for testing.  Next, try compiling a simple "hello world" file, and tell me what the output from gcc is.

**Comment by bjodah** [comment]
Commented out the tests and as per your instruction:
```
bjorn@urania:~$ conda create -n testing gcc
Fetching package metadata: ......
Solving package specifications: .
Package plan for installation in environment /home/bjorn/miniconda/envs/testing:

The following NEW packages will be INSTALLED:

    cloog: 0.18.0-0 asmeurer
    gcc:   4.8.5-0  asmeurer
    gmp:   5.1.2-2  asmeurer
    isl:   0.12.2-0 asmeurer
    mpc:   1.0.1-0  asmeurer
    mpfr:  3.1.2-0  asmeurer

Proceed ([y]/n)? y

Linking packages ...
[      COMPLETE      ]|########################################################################################################################################################################################| 100%
#
# To activate this environment, use:
# $ source activate testing
#
# To deactivate this environment, use:
# $ source deactivate
#
bjorn@urania:~$ source ~/miniconda/bin/activate testing
discarding /home/bjorn/miniconda/bin from PATH
prepending /home/bjorn/miniconda/envs/testing/bin to PATH
(testing)bjorn@urania:~$ cd tmp
(testing)bjorn@urania:~/tmp$ echo -e '#include <stdio.h>\nint main() { printf("Hello world!"); return 0; }'>a.c
(testing)bjorn@urania:~/tmp$ gcc a.c
gcc: fatal error: braced spec ‘%:sanitize(address):%{!shared:libasan_preinit%O%s} %{static-libasan:%{!shared:-Bstatic --whole-archive -lasan --no-whole-archive -Bdynamic}}%{!static-libasan:-lasan}}     %{%:sanitize(thread):%{static-libtsan:%{!shared:-Bstatic --whole-archive -ltsan --no-whole-archive -Bdynamic}}%{!static-libtsan:-ltsan}}     %{%:sanitize(leak):%{static-liblsan:%{!shared:-Bstatic --whole-archive -llsan --no-whole-archive -Bdynamic}}%{!static-liblsan:-llsan}}’ is invalid at ‘%’
compilation terminated.
```
So the error message is essentially the same.

**Comment by stuarteberg** [comment]
OK, thanks for the update.  Would you mind trying yet another version of the package?  Please try the `stuarteberg` channel instead of `asmeurer`:

    conda install -c stuarteberg gcc=4.8.5

**Comment by stuarteberg** [comment]
BTW, Just to be clear: I appreciate your help in sorting out this issue with the gcc package.  But if you need to move on with your work immediately, I think your original question was on the right track: if you just remove the gcc dependency from `gsl/meta.yaml`, you can probably build the recipe with your own version of gcc.

**Comment by bjodah** [comment]
No, I am not in a hurry, I am just trying to provide a working conda recipe for a package providing python bindings for gsl_odeiv2 (https://github.com/bjodah/pygslodeiv2), and thought I should report the issue. All your help is much appreciated.

Your build does work:
```
bjorn@urania:~$ source activate testing
discarding /home/bjorn/miniconda/bin from PATH
prepending /home/bjorn/miniconda/envs/testing/bin to PATH
(testing)bjorn@urania:~$ conda install -c stuarteberg gcc=4.8.5
Fetching package metadata: ........
Solving package specifications: .
Package plan for installation in environment /home/bjorn/miniconda/envs/testing:

The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    gcc-4.8.5.99               |                2       252.8 MB  stuarteberg

The following packages will be UPDATED:

    gcc: 4.8.5-0 asmeurer --> 4.8.5.99-2 stuarteberg

Proceed ([y]/n)? 

Fetching packages ...
gcc-4.8.5.99-2 100% |##########################################################| Time: 0:00:29   8.86 MB/s
Extracting packages ...
[      COMPLETE      ]|#############################################################################| 100%
Unlinking packages ...
[      COMPLETE      ]|#############################################################################| 100%
Linking packages ...
[      COMPLETE      ]|#############################################################################| 100%
(testing)bjorn@urania:~$ cd tmp
(testing)bjorn@urania:~/tmp$ gcc a.c
(testing)bjorn@urania:~/tmp$ ./a.out 
Hello world!(testing)bjorn@urania:~/tmp$ which gcc
/home/bjorn/miniconda/envs/testing/bin/gcc
(testing)bjorn@urania:~/tmp$
```
Do you know what made the difference? If you want help bisecting I'll gladly help out.

**Comment by stuarteberg** [comment]
>Do you know what made the difference?

Since you're on Ubuntu, and your error message complained about the syntax of the `specs` file, I made [this commit](https://github.com/conda/conda-recipes/commit/4aeb941bce446af5397fa74db3e3d9cab7c9e6ae) yesterday.  (The details aren't interesting.)  Thanks for the bug report.

Fortunately, that commit is part of a larger [PR](https://github.com/conda/conda-recipes/pull/401) that was merged earlier today.  The build I gave you included the fix, it isn't an official build.  The "official" gcc package won't be available until today's changes percolate into a new package on the conda `defaults` channel.
